source/lac/cuda_vector.cu

```diff
@@ -0,0 +1,1006 @@
+                                                           result_device, val,
+                                                           n_elements);
+
+      // Copy the result back to the host
+      Number result;
+      error_code = cudaMemcpy(&result, result_device, sizeof(Number),
+                              cudaMemcpyDeviceToHost);
+      CudaAssert(error_code);
+      // Free the memory on the device
+      error_code = cudaFree(result_device);
+      CudaAssert(error_code);
+
+      return result;
+    }
+
+
+
+    template <typename Number>
+    Number Vector<Number>::add_and_dot(const Number                     a,
+                                       const VectorSpaceVector<Number> &V,
+                                       const VectorSpaceVector<Number> &W)
+    {
+      AssertIsFinite(a);
+
+      // Check that casting will work
+      Assert(dynamic_cast<const Vector<Number>*>(&V)!=nullptr,
+             ExcVectorTypeNotCompatible());
+      Assert(dynamic_cast<const Vector<Number>*>(&W)!=nullptr,
+             ExcVectorTypeNotCompatible());
+
+      // Downcast V and W. If it fails, throw an exceptiion.
+      const Vector<Number> &down_V = dynamic_cast<const Vector<Number>&>(V);
+      Assert(down_V.size() == this->size(),
+             ExcMessage("Vector V has the wrong size."));
+      const Vector<Number> &down_W = dynamic_cast<const Vector<Number>&>(W);
+      Assert(down_W.size() == this->size(),
+             ExcMessage("Vector W has the wrong size."));
+
+      Number *res_d;
+      cudaError_t error_code = cudaMalloc(&res_d, sizeof(Number));
+      CudaAssert(error_code);
+      error_code = cudaMemset(res_d, 0., sizeof(Number));
+      CudaAssert(error_code);
+
+      const int n_blocks = 1 + (n_elements-1)/(VR_CHUNK_SIZE*VR_BKSIZE);
+      internal::add_and_dot<Number> <<<dim3(n_blocks,1),dim3(VR_BKSIZE)>>>(
+        res_d, val, down_V.val, down_W.val, a, n_elements);
+
+      Number res;
+      error_code = cudaMemcpy(&res, res_d, sizeof(Number), cudaMemcpyDeviceToHost);
+      CudaAssert(error_code);
+      error_code = cudaFree(res_d);
+
+      return res;
+    }
+
+
+
+    template <typename Number>
+    void Vector<Number>::print(std::ostream       &out,
+                               const unsigned int  precision,
+                               const bool          scientific,
+                               const bool          across) const
+    {
+      AssertThrow(out, ExcIO());
+      std::ios::fmtflags old_flags = out.flags();
+      unsigned int old_precision = out.precision (precision);
+
+      out.precision (precision);
+      if (scientific)
+        out.setf (std::ios::scientific, std::ios::floatfield);
+      else
+        out.setf (std::ios::fixed, std::ios::floatfield);
+
+      out << "IndexSet: ";
+      complete_index_set(n_elements).print(out);
+      out << std::endl;
+      for (unsigned int i=0; i<n_elements; ++i)
+        out << val[i] << std::endl;
```

Comment:
val[] is on the GPU, we should *not* be able to access it from the host like this. I guess this could work with modern CUDA Managed memory, but I think we should do an explicit copy to host and then print. Right?